zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2803,124 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a unordered list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
+            return childrenPaginated;
+        }
+
+        Set<String> childrenSet = new HashSet<>(childrenPaginated);
+        getChildrenRemainingPages(path, watcher, nextPage, childrenSet);
+
+        return new ArrayList<>(childrenSet);
+    }
+
+    private void getChildrenRemainingPages(String path,
+                                           Watcher watcher,
+                                           PaginationNextPage nextPage,
+                                           Set<String> childrenSet) throws KeeperException, InterruptedException {
+        int retryCount = 0;
+        do {
+            List<String> childrenPaginated;
+            long minCzxid = nextPage.getMinCzxid();
+            if (nextPage.getMinCzxidOffset() == 0) {
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
+                        nextPage.getMinCzxidOffset(), null, nextPage);
+            } else {
+                // If a child with the same czxid is returned in the previous page, and then deleted
+                // on the server, the children not in the previous page but offset is less than czxidOffset
+                // would be missed in the next page. Use the last child in the previous page to determine whether or not
+                // the deletion case happens. If it happens, retry fetching the page starting from offset 0.
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid,
+                        nextPage.getMinCzxidOffset() - 1, null, nextPage);
+                if (!childrenPaginated.isEmpty() && !childrenSet.contains(childrenPaginated.get(0))) {
+                    if (retryCount++ >= 3) {
+                        throw KeeperException.create(Code.OPERATIONTIMEOUT);
+                    }
+                    LOG.info("Retry fetching children for minZxid = {}, retries = {}", minCzxid, retryCount);
+                    childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid, 0, null, nextPage);
```

Comment:
Also, here, do we need to clear `childrenSet` here? As retry from beginning, we will test against no empty `childrenSet` from the previous run. This may give you false result even deletion happens again?

Reply:
Shall we have a test case to cover it?

Reply:
Addressed. I thought about the test. I can't find a way to test such case that a deletion happens between two pages as it automatically fetches next page here. I can't insert a countdown latch for a deletion to happen before the next page. Please advise if you have an idea.